File: ios/Zest/ViewModels/CookbooksViewModel.swift
import Foundation
import FirebaseAuth

@MainActor
class CookbooksViewModel: ObservableObject {
    @Published var cookbooks: [Cookbook] = []
    @Published var isLoading = false
    @Published var errorMessage: String?
    
    private let firebaseService = FirebaseService.shared
    
    func loadCookbooks() async {
        guard let userId = Auth.auth().currentUser?.uid else {
            errorMessage = "User not authenticated"
            return
        }
        
        isLoading = true
        errorMessage = nil
        
        do {
            cookbooks = try await firebaseService.getCookbooks(userId: userId)
        } catch {
            errorMessage = error.localizedDescription
        }
        
        isLoading = false
    }
    
    func createCookbook(name: String, description: String?, isPublic: Bool = false) async -> Bool {
        guard let userId = Auth.auth().currentUser?.uid else {
            errorMessage = "User not authenticated"
            return false
        }
        
        let cookbook = Cookbook(
            name: name,
            description: description,
            recipeIds: [],
            userId: userId,
            isPublic: isPublic,
            createdAt: Date(),
            updatedAt: Date()
        )
        
        do {
            let _ = try await firebaseService.createCookbook(cookbook)
            await loadCookbooks()
            return true
        } catch {
            errorMessage = error.localizedDescription
            return false
        }
    }
    
    func updateCookbook(id: String, name: String, description: String?, isPublic: Bool) async -> Bool {
        guard let index = cookbooks.firstIndex(where: { $0.id == id }) else {
            errorMessage = "Cookbook not found"
            return false
        }
        
        var cookbook = cookbooks[index]
        cookbook.name = name
        cookbook.description = description
        cookbook.isPublic = isPublic
        cookbook.updatedAt = Date()
        
        do {
            try await firebaseService.updateCookbook(id: id, cookbook)
            await loadCookbooks()
            return true
        } catch {
            errorMessage = error.localizedDescription
            return false
        }
    }
    
    func deleteCookbook(id: String) async -> Bool {
        do {
            try await firebaseService.deleteCookbook(id: id)
            await loadCookbooks()
            return true
        } catch {
            errorMessage = error.localizedDescription
            return false
        }
    }
    
    func addRecipeToCookbook(cookbookId: String, recipeId: String) async -> Bool {
        do {
            try await firebaseService.addRecipeToCookbook(cookbookId: cookbookId, recipeId: recipeId)
            await loadCookbooks()
            return true
        } catch {
            errorMessage = error.localizedDescription
            return false
        }
    }
    
    func removeRecipeFromCookbook(cookbookId: String, recipeId: String) async -> Bool {
        do {
            try await firebaseService.removeRecipeFromCookbook(cookbookId: cookbookId, recipeId: recipeId)
            await loadCookbooks()
            return true
        } catch {
            errorMessage = error.localizedDescription
            return false
        }
    }
}

File: android/app/src/main/java/com/neostudios/zest/ui/screens/CookbooksViewModel.kt
package com.neostudios.zest.ui.screens

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.neostudios.zest.domain.model.Cookbook
import com.neostudios.zest.domain.model.Recipe
import com.neostudios.zest.data.repository.CookbookRepository
import com.neostudios.zest.data.repository.RecipeRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class CookbooksViewModel @Inject constructor(
    private val cookbookRepository: CookbookRepository,
    private val recipeRepository: RecipeRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(CookbooksUiState())
    val uiState: StateFlow<CookbooksUiState> = _uiState

    init {
        loadCookbooks()
        loadAllRecipes()
    }

    fun loadCookbooks() {
        _uiState.value = _uiState.value.copy(isLoading = true)
        viewModelScope.launch {
            try {
                val cookbooks = cookbookRepository.getCookbooks()
                _uiState.value = _uiState.value.copy(
                    cookbooks = cookbooks,
                    isLoading = false
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    error = e.localizedMessage,
                    isLoading = false
                )
            }
        }
    }

    private fun loadAllRecipes() {
        viewModelScope.launch {
            try {
                val recipes = recipeRepository.getAllRecipes()
                _uiState.value = _uiState.value.copy(allRecipes = recipes)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.localizedMessage)
            }
        }
    }

    fun createCookbook(name: String, description: String, isPublic: Boolean) {
        viewModelScope.launch {
            try {
                val newCookbook = Cookbook(
                    name = name,
                    description = description,
                    isPublic = isPublic,
                    recipeIds = emptyList()
                )
                cookbookRepository.createCookbook(newCookbook)
                loadCookbooks()
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.localizedMessage)
            }
        }
    }

    fun updateCookbook(cookbook: Cookbook) {
        viewModelScope.launch {
            try {
                cookbookRepository.updateCookbook(cookbook)
                loadCookbooks()
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.localizedMessage)
            }
        }
    }

    fun deleteCookbook(cookbookId: String) {
        viewModelScope.launch {
            try {
                cookbookRepository.deleteCookbook(cookbookId)
                loadCookbooks()
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.localizedMessage)
            }
        }
    }

    fun addRecipeToCookbook(cookbookId: String, recipeId: String) {
        viewModelScope.launch {
            try {
                val cookbook = _uiState.value.cookbooks.find { it.id == cookbookId }
                if (cookbook != null && !cookbook.recipeIds.contains(recipeId)) {
                    val updatedRecipeIds = cookbook.recipeIds + recipeId
                    cookbookRepository.updateCookbook(cookbook.copy(recipeIds = updatedRecipeIds))
                    loadCookbooks()
                }
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.localizedMessage)
            }
        }
    }

    fun removeRecipeFromCookbook(cookbookId: String, recipeId: String) {
        viewModelScope.launch {
            try {
                val cookbook = _uiState.value.cookbooks.find { it.id == cookbookId }
                if (cookbook != null && cookbook.recipeIds.contains(recipeId)) {
                    val updatedRecipeIds = cookbook.recipeIds.filter { it != recipeId }
                    cookbookRepository.updateCookbook(cookbook.copy(recipeIds = updatedRecipeIds))
                    loadCookbooks()
                }
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.localizedMessage)
            }
        }
    }
}

data class CookbooksUiState(
    val cookbooks: List<Cookbook> = emptyList(),
    val allRecipes: List<Recipe> = emptyList(),
    val isLoading: Boolean = false,
    val error: String? = null
)
